File: day6/day6.scala
import io.Source
import scala.util.Using
import scala.collection.mutable.ListBuffer
import scala.collection.Seq

def readProblems(filename: String): List[List[String]] =
  val content = Source.fromFile(filename).getLines().toList.reverse
  val tmp = content.head
    .toCharArray()
    .zipWithIndex
    .filterNot(_._1 == ' ')
    .map(_._2)
    .appended(content.head.length())
    .toList

  tmp
    .dropRight(1)
    .zip(tmp.drop(1))
    .map((l, h) => content.map(_.substring(l, h)))
    .toList

def solve_part_1(problems: List[List[String]]): Long =
  problems
    .map(problem => {
      val tmp = problem.tail.map(_.trim).filterNot(_.isEmpty()).map(_.toLong)
      if problem.head.trim() == "*" then tmp.product else tmp.sum
    })
    .sum

def solve_part_2(problems: List[List[String]]): Long =
  solve_part_1(
    problems
      .map(problem =>
        problem.head :: problem.tail.reverse
          .map(_.toCharArray())
          .transpose
          .map(_.mkString)
      )
  )

@main def main(filename: String) = {
  val problems = readProblems(filename)
  println(solve_part_1(problems))
  println(solve_part_2(problems))
}

File: day6/day6.exs
defmodule Day6 do
  def read_problems(filename) do
    content =
      File.stream!(filename)
      |> Enum.reverse()
      |> Enum.to_list()

    content
    |> List.first()
    |> then(&String.graphemes/1)
    |> Enum.with_index()
    |> Enum.reject(&(elem(&1, 0) == " "))
    |> Enum.map(&elem(&1, 1))
    |> then(&Enum.zip(Enum.drop(&1, -1), Enum.drop(&1, 1)))
    |> Enum.map(fn {l, h} ->
      content |> Enum.map(&String.slice(&1, l..(h - 1)))
    end)
  end

  def solve_part_1(problems) do
    problems
    |> Enum.map(fn [op | operands] ->
      f = if String.trim(op) == "*", do: &Enum.product/1, else: &Enum.sum/1

      operands
      |> Enum.map(&String.trim/1)
      |> Enum.reject(&(&1 == ""))
      |> Enum.map(&String.to_integer/1)
      |> then(&f.(&1))
    end)
    |> Enum.sum()
  end

  def solve_part_2(problems) do
    problems
    |> Enum.map(fn [hd | tl] ->
      [
        hd
        | tl
          |> Enum.reverse()
          |> Enum.map(&String.split(&1, ""))
          |> Enum.zip()
          |> Enum.map(&Tuple.to_list/1)
          |> Enum.map(&Enum.join/1)
      ]
    end)
    |> then(&solve_part_1/1)
  end
end

filename = System.argv() |> List.first()
problems = Day6.read_problems(filename)
Day6.solve_part_1(problems) |> IO.puts()
Day6.solve_part_2(problems) |> IO.puts()
